design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,205 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a custer-level rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object without interupting the workload.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to eventually propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out ReplicaSets.
+
+
+## Goals
+
+-   Provide a way to eventually propagate changes from a running VM object to its corresponding VMI without interrupting the workload.
+    
+-   Maintain backwards compatibility for cluster owners that prefer keeping staged changes until next VM reboot.
+    
+
+## Non Goals
+
+-   Live uppdate for **all** fields of the Virtual Machine API. Only select fields can be propagated to the guest. 
+-   Make LiveUpdate the default. We defer this decision.
+-   Support upgrades of VMs that are using the `liveUpdateFeatures`.
+    
+
+## Definition of Users
+
+A VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a VM owner I want changes that I applied to the VM object to propagate to the VMI if possible.
+
+As a VM owner, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+> **Note:** Below we are only discussing changes to fields that support live updates.
+                  References to `all` fields only refers to fields that can be live-updated. 
+
+## KubeVirt CR
+
+A new field called `vmRolloutStrategy` will be introduced in the KubeVirt CR. This field will control the behavior of the VM objects. Two possible values will be available:
+
+1. `LiveUpdate` means all changes to the VM object will be eventually applied to the VMI.
+    
+2. `Stage` would mean the current familiar behavior remains - all changes to the VM object will be staged and applied after a restart.
+
+In both cases, a condition called `RestartRequired` will be added to the VM to reflect the need for a restart and the reason why.
+
+The default value of the `rolloutStrategy` field if not specified will be `Stage`.
+
+The `vmRolloutStrategy` field will be usable only if the `VMLiveUpdateFeatures` feature gate is enabled.
+
+
+## VM API
+
+The VM level `liveUpdateFeatures` field will be removed.
+VMs using the `liveUpdateFeatures` will be rejected by the API.
+
+If `RestartRequired` condition is present, any additional changes to the VM spec - to live-updatedable or non-liveupdateable fields - will be staged. Thus the VM has to be restarted.
```

Comment:
Can you explain this?
This condition can appear because a change that requires restart is set, but a different change after may not require a restart and perhaps is applied immediately.

The way it is defined here, sounds as if the following changes, even if they can be applied immediate, will not.

Reply:
> Well, this will cause a complication either way.

I'm missing the "why"? :)

> I would prefer for everything to do its best to reach "sync".

I'm not seeing anything which would prevent to enhance/optimize this further in future.

Reply:
> > Well, this will cause a complication either way.
>
> I'm missing the "why"? :)

From the user perspective: It matters which field he changed first, that will determine if the whole future changes will take effect immediately or not
From the implementation: On any change detected, there is a need to check the status and examine the condition. Therefore, not only the fields that changes determines if the change should role out to the VMI, but also its condition status. There are corner cases with this approach, as the status may or may not be up to date, dependent on who sets the condition state and when it is read.
Perhaps this is much simpler then what I imagine.

> > I would prefer for everything to do its best to reach "sync".
>
> I'm not seeing anything which would prevent to enhance/optimize this further in future.

Correct, but I am also not seeing a problem to do the best possible to sync changes based on the available information. The fact that some prev change requires reset is not a reason a new change cannot be applied. In the end, we do have the desired state in the spec, if there are dependencies they can be checked against the desired and current.

Anyway, as it can be improved, I guess it is not critical at the moment.
It will be critical if you consider totally different areas being added to this rollout logic, e.g. storage, virt infra or network being dependent on one another in this regard.

Reply:
From a user perspective I agree with Ed.
Maybe I did a change that requires a restart (say add memory for now) and then I am adding a NIC, because I need to deploy some software over that network. Why would I not get this NIC hotplugged?

I think this will cause quite some confusion.
If the change cannot be applied, I think it would be better to issue a Warning/and Event telling that the hot-change failed.

Reply:
`RestartRequired` means the VM object is now out of sync with its VMI.
From that point on, the VM should not be allowed to influence the VMI until the disconnect is resolved (by rebooting).
Allowing live updates when `RestartRequired` is present requires very complex logic, which would be a lot more confusing to the user than just "Can't make new live changes until next restart, sorry".

For example, take a VM with 1 `sockets` and 4 `maxSockets` and **either**:
- Change `maxSockets` to 2, there's now a `restartRequired` on the VM. Now, do you allow updating `sockets` to 2? What about 3? Technically, `sockets` can be live updated to 3, but that would lead to an invalid VM spec...
- Change `maxSockets` to 6, now you can technically allow live-updating `sockets` up to 4, but you need to remember the old value of `maxSockets` (how?). If `sockets` is set to 5 that needs to be staged...

A partial solution to this would be to stage all CPU-related changes as soon as a CPU-related change triggers a `RestartRequired`, but how do we link the condition to a category of things? What about incompatibilities between sections? (for example, CPUs have a limit on how much memory they can handle, to go above you need more sockets)

For user's sake and implementation sake, I really think `RestartRequired` should be incompatible with live updates. As mentioned in previous comments, if in the future we find safe ways to be more fine-grained, we can always improve the logic!

Reply:
> For user's sake and implementation sake, I really think `RestartRequired` should be incompatible with live updates. As mentioned in previous comments, if in the future we find safe ways to be more fine-grained, we can always improve the logic!

+1 I agree.

I don't see how we can commit to being able to process conflicting changes on the VM spec in one iteration. The following proposal can undo this decision.

Reply:
Thanks for the example.

I think that the networking side is coming from a side with simple examples:
1. modify sockets --> requires restart
2. add vnic and network, no conflict, no error, works

For my part I don't know how often this will work, but my point is that there are _some_ orthorgonal changes wich :migh_ often work to be applied without impacting the currently staged changes.

But maybe we can separate the discussion about the semantics of `RestartRequired` and the tradeoffs around it

If we seaparate it, then I'd propose to reword the sentence to: 

   If `RestartRequired` condition is present, the VM has to be restarted.